game/script-snap.rpy

```diff
@@ -0,0 +1,778 @@
+                _cards_on_table.append(new_card)
+                renpy.play("mod_assets/sfx/card_place.mp3")
+                _is_player_turn = False
+
+        else:
+            if (len(_natsuki_hand) > 0):
+                new_card = _natsuki_hand.pop(0)
+                _cards_on_table.append(new_card)
+                renpy.play("mod_assets/sfx/card_place.mp3")
+                _is_player_turn = True
+
+        update_turn_indicator()
+        draw_card_onscreen()
+
+    def _get_card_label_to_display():
+        """
+        Returns a string representing the uppermost card on the table pile
+        """
+        if len(_cards_on_table) >= 1:
+            return "{0} of {1}".format(_cards_on_table[-1][0], _cards_on_table[-1][1])
+
+        else:
+            return "None!"
+
+    def _get_snap_result():
+        """
+        Compares the last two cards placed on the table pile, and returns True if either:
+            - The suits on both cards match
+            - The values on both cards match
+        Otherwise, returns False
+        Used by Natsuki's logic to determine if she should "spot" the snap opportunity
+        """
+        if len(_cards_on_table) >= 2:
+            return _cards_on_table[-1][0] == _cards_on_table[-2][0] or _cards_on_table[-1][1] == _cards_on_table[-2][1]
+        
+        else:
+            return False
+
+    def _call_snap(is_player=False):
+        """
+        Attempts to call snap and award cards for the player or Natsuki, based on who made the call
+
+        IN:
+            - is_player boolean value representing if the player or Natsuki was the one who made the call
+        """
+        global _is_player_turn
+        global _player_is_snapping
+        global _snap_popup_image
+
+        # We set this here so Natsuki can't try to snap while the player is snapping
+        if is_player:
+            _player_is_snapping = True
+
+        # If the suit/value on the last placed card matches the preceding card, the snap is valid
+        if _get_snap_result():
+
+            if is_player:
+                # Player called snap successfully; give them the cards on the table
+                for card in _cards_on_table:
+                    _player_hand.append(card)
+
+            else:
+                # Natsuki called snap successfully; give her the cards on the table
+                for card in _cards_on_table:
+                    _natsuki_hand.append(card)
+
+            # Clear the cards on the table
+            del _cards_on_table[:]
+            renpy.play("mod_assets/sfx/card_shuffle.mp3")
+
+            _snap_popup_image = random.choice(_snap_popup_sprites)
+            draw_card_onscreen()
+
+            # Natsuki comments on the correct snap
+            renpy.call("snap_quip", is_player_snap=is_player, is_correct_snap=True)
+
+        else:
+            # Natsuki comments on the incorrect snap
+            renpy.call("snap_quip", is_player_snap=is_player, is_correct_snap=False)
+
```

Comment:
Is there any reason you're using renpy.call over doing all the logic here and calling the `renpy.say` line w/ a substitute?

renpy.call is extremely janky and we should avoid using it wherever we possibly can due to how it breaks out of all `python` blocks.

Using this often would make it seem acceptable for contributors to use it elsewhere and stumble into the same issue and wonder why it's not working.

Arguably not our fault, but we should avoid perpetuating a use for it.

Reply:
Aaaaaaaargh, I remember this being a massive headache.

The original approach did use `jump`, but from memory had some very annoying issues with flow - it's been a long time since I looked at this and realistically I'd want to tackle this in its own dedicated feature/bugfix rather than prolong test/blizz any longer.

Suggest retention as a stop-gap for purpose of knocking this branch over the head, to be revisited along with other reworks we'll need pre-release. Will add a comment advising against this approach to dissuade replication of approach until then.